In [1]:
import pandas as pd

In [2]:
basket_ball_team = pd.read_csv("nba.csv")

In [3]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [4]:
basket_ball_team.tail()

Name       Team  Number Position   Age Height  Weight College  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   

        Salary  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0  
457        NaN

# SHARED METHODS AND ATTRBUTES

In [5]:
basket_ball_team.index

RangeIndex(start=0, stop=458, step=1)

In [6]:
basket_ball_team.values

array([['Avery Bradley', 'Boston Celtics', 0.0, ..., 180.0, 'Texas',
        7730337.0],
       ['Jae Crowder', 'Boston Celtics', 99.0, ..., 235.0, 'Marquette',
        6796117.0],
       ['John Holland', 'Boston Celtics', 30.0, ..., 205.0,
        'Boston University', nan],
       ...,
       ['Tibor Pleiss', 'Utah Jazz', 21.0, ..., 256.0, nan, 2900000.0],
       ['Jeff Withey', 'Utah Jazz', 24.0, ..., 231.0, 'Kansas', 947276.0],
       [nan, nan, nan, ..., nan, nan, nan]], dtype=object)

In [7]:
basket_ball_team.shape

(458, 9)

# shape tells = rows*colums

In [8]:
basket_ball_team.dtypes

Name         object
Team         object
Number      float64
Position     object
Age         float64
Height       object
Weight      float64
College      object
Salary      float64
dtype: object

In [9]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [10]:
basket_ball_team.columns

Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
       'College', 'Salary'],
      dtype='object')

In [11]:
basket_ball_team.index

RangeIndex(start=0, stop=458, step=1)

In [12]:
basket_ball_team.axes

[RangeIndex(start=0, stop=458, step=1),
 Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
        'College', 'Salary'],
       dtype='object')]

# the code above bundles those two together in single output

In [13]:
basket_ball_team.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     373 non-null object
Salary      446 non-null float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


# object means strings and float is floating number

In [14]:
basket_ball_team.get_dtype_counts()

float64    4
object     5
dtype: int64

# DIFFERECES BETWEEN SHARED METHODS

In [15]:
pd.read_csv("revenue.csv")

Date  New York  Los Angeles  Miami
0   1/1/16       985          122    499
1   1/2/16       738          788    534
2   1/3/16        14           20    933
3   1/4/16       730          904    885
4   1/5/16       114           71    253
5   1/6/16       936          502    497
6   1/7/16       123          996    115
7   1/8/16       935          492    886
8   1/9/16       846          954    823
9  1/10/16        54          285    216

In [16]:
revenue = pd.read_csv("revenue.csv", index_col = "Date")

In [17]:
revenue

New York  Los Angeles  Miami
Date                                 
1/1/16        985          122    499
1/2/16        738          788    534
1/3/16         14           20    933
1/4/16        730          904    885
1/5/16        114           71    253
1/6/16        936          502    497
1/7/16        123          996    115
1/8/16        935          492    886
1/9/16        846          954    823
1/10/16        54          285    216

In [18]:
revenue.sum()

New York       5475
Los Angeles    5134
Miami          5641
dtype: int64

# the two codes below will give the same result as the code above

In [19]:
revenue.sum(axis = "index")

New York       5475
Los Angeles    5134
Miami          5641
dtype: int64

In [20]:
revenue.sum(axis = 0)

New York       5475
Los Angeles    5134
Miami          5641
dtype: int64

# above code gave sum in vertical order - for each city, if you want the sum horizontally - ie: by date, use the code below

In [21]:
revenue.sum(axis = 1)

Date
1/1/16     1606
1/2/16     2060
1/3/16      967
1/4/16     2519
1/5/16      438
1/6/16     1935
1/7/16     1234
1/8/16     2313
1/9/16     2623
1/10/16     555
dtype: int64

# now the code gave the sum = date wise, ie: horizontally=> it added the data for each date

In [22]:
revenue.sum(axis="columns")

Date
1/1/16     1606
1/2/16     2060
1/3/16      967
1/4/16     2519
1/5/16      438
1/6/16     1935
1/7/16     1234
1/8/16     2313
1/9/16     2623
1/10/16     555
dtype: int64

In [23]:
revenue.sum(axis= 1)

Date
1/1/16     1606
1/2/16     2060
1/3/16      967
1/4/16     2519
1/5/16      438
1/6/16     1935
1/7/16     1234
1/8/16     2313
1/9/16     2623
1/10/16     555
dtype: int64

# SELECT ONE COLUMN FROM A DATA FRAME

In [24]:
basket_ball_team = pd.read_csv("nba.csv")

In [25]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# when you extract a single column from a data frame it just gives you a series 

In [26]:
basket_ball_team.Name

0                Avery Bradley
1                  Jae Crowder
2                 John Holland
3                  R.J. Hunter
4                Jonas Jerebko
5                 Amir Johnson
6                Jordan Mickey
7                 Kelly Olynyk
8                 Terry Rozier
9                 Marcus Smart
10             Jared Sullinger
11               Isaiah Thomas
12                 Evan Turner
13                 James Young
14                Tyler Zeller
15            Bojan Bogdanovic
16                Markel Brown
17             Wayne Ellington
18     Rondae Hollis-Jefferson
19                Jarrett Jack
20              Sergey Karasev
21             Sean Kilpatrick
22                Shane Larkin
23                 Brook Lopez
24            Chris McCullough
25                 Willie Reed
26             Thomas Robinson
27                  Henry Sims
28                Donald Sloan
29              Thaddeus Young
                ...           
428            Al-Farouq Aminu
429     

In [27]:
basket_ball_team.Team

0              Boston Celtics
1              Boston Celtics
2              Boston Celtics
3              Boston Celtics
4              Boston Celtics
5              Boston Celtics
6              Boston Celtics
7              Boston Celtics
8              Boston Celtics
9              Boston Celtics
10             Boston Celtics
11             Boston Celtics
12             Boston Celtics
13             Boston Celtics
14             Boston Celtics
15              Brooklyn Nets
16              Brooklyn Nets
17              Brooklyn Nets
18              Brooklyn Nets
19              Brooklyn Nets
20              Brooklyn Nets
21              Brooklyn Nets
22              Brooklyn Nets
23              Brooklyn Nets
24              Brooklyn Nets
25              Brooklyn Nets
26              Brooklyn Nets
27              Brooklyn Nets
28              Brooklyn Nets
29              Brooklyn Nets
                ...          
428    Portland Trail Blazers
429    Portland Trail Blazers
430    Por

# if there are spaces in column name the above method does not work 100%, so the best way to extract single column is the code below

In [28]:
basket_ball_team["Name"]

0                Avery Bradley
1                  Jae Crowder
2                 John Holland
3                  R.J. Hunter
4                Jonas Jerebko
5                 Amir Johnson
6                Jordan Mickey
7                 Kelly Olynyk
8                 Terry Rozier
9                 Marcus Smart
10             Jared Sullinger
11               Isaiah Thomas
12                 Evan Turner
13                 James Young
14                Tyler Zeller
15            Bojan Bogdanovic
16                Markel Brown
17             Wayne Ellington
18     Rondae Hollis-Jefferson
19                Jarrett Jack
20              Sergey Karasev
21             Sean Kilpatrick
22                Shane Larkin
23                 Brook Lopez
24            Chris McCullough
25                 Willie Reed
26             Thomas Robinson
27                  Henry Sims
28                Donald Sloan
29              Thaddeus Young
                ...           
428            Al-Farouq Aminu
429     

In [29]:
basket_ball_team["College"]

0                  Texas
1              Marquette
2      Boston University
3          Georgia State
4                    NaN
5                    NaN
6                    LSU
7                Gonzaga
8             Louisville
9         Oklahoma State
10            Ohio State
11            Washington
12            Ohio State
13              Kentucky
14        North Carolina
15                   NaN
16        Oklahoma State
17        North Carolina
18               Arizona
19          Georgia Tech
20                   NaN
21            Cincinnati
22            Miami (FL)
23              Stanford
24              Syracuse
25           Saint Louis
26                Kansas
27            Georgetown
28             Texas A&M
29          Georgia Tech
             ...        
428          Wake Forest
429           Notre Dame
430           California
431       North Carolina
432           St. John's
433                 Duke
434     Central Michigan
435             Illinois
436          Weber State


In [30]:
type(basket_ball_team["College"])

pandas.core.series.Series

# extracting single column always gives us a series

In [31]:
basket_ball_team["College"].head()

0                Texas
1            Marquette
2    Boston University
3        Georgia State
4                  NaN
Name: College, dtype: object

In [32]:
basket_ball_team =pd.read_csv("nba.csv")

In [33]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# EXTRACT TWO OR MORE COLUMNS FROM DATA FRAME

# use a list within square brackets to extract the columns

In [34]:
basket_ball_team[ ["Name", "College", "Team"]  ].head()

Name            College            Team
0  Avery Bradley              Texas  Boston Celtics
1    Jae Crowder          Marquette  Boston Celtics
2   John Holland  Boston University  Boston Celtics
3    R.J. Hunter      Georgia State  Boston Celtics
4  Jonas Jerebko                NaN  Boston Celtics

In [35]:
basket_ball_team[["Team", "Position"]].head()

Team Position
0  Boston Celtics       PG
1  Boston Celtics       SF
2  Boston Celtics       SG
3  Boston Celtics       SG
4  Boston Celtics       PF

In [36]:
needed_columns = ["Name","Team","Position"]

In [37]:
basket_ball_team[needed_columns].head()

Name            Team Position
0  Avery Bradley  Boston Celtics       PG
1    Jae Crowder  Boston Celtics       SF
2   John Holland  Boston Celtics       SG
3    R.J. Hunter  Boston Celtics       SG
4  Jonas Jerebko  Boston Celtics       PF

# ADDING NEW COLUMN TO DATA FRAME

In [38]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [39]:
basket_ball_team["Sport"] = "Basket Ball"

In [40]:
basket_ball_team.head(3)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   

             College     Salary        Sport  
0              Texas  7730337.0  Basket Ball  
1          Marquette  6796117.0  Basket Ball  
2  Boston University        NaN  Basket Ball

# new column sport is added to the table

In [43]:
basket_ball_team["League"] = "National Basketball Association"

In [44]:
basket_ball_team.head(3)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   

             College     Salary        Sport                           League  
0              Texas  7730337.0  Basket Ball  National Basketball Association  
1          Marquette  6796117.0  Basket Ball  National Basketball Association  
2  Boston University        NaN  Basket Ball  National Basketball Association

In [45]:
basket_ball_team = pd.read_csv("nba.csv")

In [47]:
basket_ball_team.head(4)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0

# INSERT METHOD TO ADD AN EXTRA COLUMN 

# LOC - index location within the columns- name is index 0, team is index 1, and so on
# LOC asks us for the numeric index position where we would like to insert the column
# column- is the name of the column
# value - is the value you want to insert in that particluar column

In [49]:
basket_ball_team.insert(3, column = "Sport", value = "Basketball")

In [50]:
basket_ball_team.head()

Name            Team  Number       Sport Position   Age Height  \
0  Avery Bradley  Boston Celtics     0.0  Basketball       PG  25.0    6-2   
1    Jae Crowder  Boston Celtics    99.0  Basketball       SF  25.0    6-6   
2   John Holland  Boston Celtics    30.0  Basketball       SG  27.0    6-5   
3    R.J. Hunter  Boston Celtics    28.0  Basketball       SG  22.0    6-5   
4  Jonas Jerebko  Boston Celtics     8.0  Basketball       PF  29.0   6-10   

   Weight            College     Salary  
0   180.0              Texas  7730337.0  
1   235.0          Marquette  6796117.0  
2   205.0  Boston University        NaN  
3   185.0      Georgia State  1148640.0  
4   231.0                NaN  5000000.0

# at index 3 the new column sport is inserted above

In [51]:
basket_ball_team.insert(7, column = "League", value = "National Basketball Association")

In [52]:
basket_ball_team.head()

Name            Team  Number       Sport Position   Age Height  \
0  Avery Bradley  Boston Celtics     0.0  Basketball       PG  25.0    6-2   
1    Jae Crowder  Boston Celtics    99.0  Basketball       SF  25.0    6-6   
2   John Holland  Boston Celtics    30.0  Basketball       SG  27.0    6-5   
3    R.J. Hunter  Boston Celtics    28.0  Basketball       SG  22.0    6-5   
4  Jonas Jerebko  Boston Celtics     8.0  Basketball       PF  29.0   6-10   

                            League  Weight            College     Salary  
0  National Basketball Association   180.0              Texas  7730337.0  
1  National Basketball Association   235.0          Marquette  6796117.0  
2  National Basketball Association   205.0  Boston University        NaN  
3  National Basketball Association   185.0      Georgia State  1148640.0  
4  National Basketball Association   231.0                NaN  5000000.0

# BROADCASTING OPERATIONS

In [53]:
basket_ball_team =pd.read_csv("nba.csv")

In [54]:
basket_ball_team.head(3)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN

In [59]:
players_age_after_5_years = basket_ball_team["Age"].add(5)

In [60]:
players_age_after_5_years.head()

0    30.0
1    30.0
2    32.0
3    27.0
4    34.0
Name: Age, dtype: float64

# added 5 years to the players age, below subtracted 10000 usd from salaries

In [62]:
salary_after_deduction = basket_ball_team["Salary"].sub(10000)

In [63]:
salary_after_deduction.head()

0    7720337.0
1    6786117.0
2          NaN
3    1138640.0
4    4990000.0
Name: Salary, dtype: float64

# you can use + 5 or -10000 in place of add or sub 

# for example: basket_ball_team["Salary"] - 10000
# or basket_ball_team["Age"] + 5

# its the same thing

# .value_counts() - great way to get a count of unique values ina single column

In [3]:
import pandas as pd

In [4]:
basket_ball_team = pd.read_csv("nba.csv")

In [5]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [9]:
basket_ball_team["Team"].value_counts()

New Orleans Pelicans      19
Memphis Grizzlies         18
Milwaukee Bucks           16
New York Knicks           16
Oklahoma City Thunder     15
Charlotte Hornets         15
Philadelphia 76ers        15
Washington Wizards        15
Sacramento Kings          15
Brooklyn Nets             15
Chicago Bulls             15
Golden State Warriors     15
Phoenix Suns              15
Indiana Pacers            15
Cleveland Cavaliers       15
Portland Trail Blazers    15
Houston Rockets           15
Toronto Raptors           15
Detroit Pistons           15
Miami Heat                15
Los Angeles Clippers      15
Dallas Mavericks          15
Los Angeles Lakers        15
San Antonio Spurs         15
Boston Celtics            15
Atlanta Hawks             15
Denver Nuggets            15
Utah Jazz                 15
Orlando Magic             14
Minnesota Timberwolves    14
Name: Team, dtype: int64

In [10]:
basket_ball_team["College"].value_counts()

Kentucky            22
Duke                20
Kansas              18
North Carolina      16
UCLA                15
Arizona             13
Florida             10
Texas                9
Syracuse             8
Washington           7
Connecticut          7
USC                  7
Michigan             6
LSU                  6
Michigan State       6
Georgetown           6
Wake Forest          6
Georgia Tech         6
Wisconsin            5
Ohio State           5
Marquette            5
Stanford             5
UNLV                 4
Oklahoma State       4
Villanova            4
Indiana              4
Tennessee            4
California           3
Utah                 3
Nevada               3
                    ..
Charleston           1
Louisiana Tech       1
Bucknell             1
Cleveland State      1
Westchester CC       1
Kansas State         1
Providence           1
Wyoming              1
Weber State          1
Tennessee State      1
Rhode Island         1
Davidson             1
Oklahoma   

In [11]:
basket_ball_team["Position"].value_counts()

SG    102
PF    100
PG     92
SF     85
C      78
Name: Position, dtype: int64

# desc order - you can use head(1) to see the most popular or common position 

In [16]:
basket_ball_team["Weight"].value_counts().tail(2)

289.0    1
307.0    1
Name: Weight, dtype: int64

In [15]:
basket_ball_team["Salary"].value_counts().head(2)

947276.0    31
845059.0    18
Name: Salary, dtype: int64

In [18]:
basket_ball_team["Height"].value_counts().head()

6-9     59
6-10    47
6-7     45
6-8     43
6-6     42
Name: Height, dtype: int64

# DROP ROWS WITH NULL VALUES

In [17]:
import pandas as pd

In [18]:
basket_ball_team = pd.read_csv("nba.csv")

In [21]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# to remove null values use the .DROPNA method

In [6]:
null_values = basket_ball_team.dropna()

In [7]:
null_values.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   

         College     Salary  
0          Texas  7730337.0  
1      Marquette  6796117.0  
3  Georgia State  1148640.0  
6            LSU  1170960.0  
7        Gonzaga  2165160.0

# it removed the row which has any column with a null value

In [10]:
complete_null_row = basket_ball_team.dropna(how='all')

In [11]:
complete_null_row.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# how = 'all' only removes rows that have an entire row with null values

# if you make inplace = True it would overwrite the existing data and give a new data set 

In [34]:
basket_ball_team.dropna(how = 'all', inplace=True)

In [42]:
basket_ball_team.tail(3)

Name       Team  Number Position   Age Height  Weight College  \
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   

        Salary  
454   900000.0  
455  2900000.0  
456   947276.0

# DROP COLUMNS WITH NULL VALUES

# axis = 1 or axis = 'columns' same thing

# how = 'any' removes any null value whereas how = 'all' removes rows or columns that all values as NULL

In [40]:
column_with_null_values = basket_ball_team.dropna(axis=1)

In [41]:
column_with_null_values.tail()

Name       Team  Number Position   Age Height  Weight
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0

In [45]:
# column college had null values at row 454 and 455, which has been removed here

# dropna has a parameter called SUBSET - you can write in which column you want it to go and just remove null values from that row 

In [51]:
salary = basket_ball_team.dropna(subset=["Salary"])

In [52]:
salary.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   

         College      Salary  
0          Texas   7730337.0  
1      Marquette   6796117.0  
3  Georgia State   1148640.0  
4            NaN   5000000.0  
5            NaN  12000000.0

# fill in values with .fillna() method

In [56]:
basket_ball_team.fillna(value=0)

Name                    Team  Number Position   Age  \
0              Avery Bradley          Boston Celtics     0.0       PG  25.0   
1                Jae Crowder          Boston Celtics    99.0       SF  25.0   
2               John Holland          Boston Celtics    30.0       SG  27.0   
3                R.J. Hunter          Boston Celtics    28.0       SG  22.0   
4              Jonas Jerebko          Boston Celtics     8.0       PF  29.0   
5               Amir Johnson          Boston Celtics    90.0       PF  29.0   
6              Jordan Mickey          Boston Celtics    55.0       PF  21.0   
7               Kelly Olynyk          Boston Celtics    41.0        C  25.0   
8               Terry Rozier          Boston Celtics    12.0       PG  22.0   
9               Marcus Smart          Boston Celtics    36.0       PG  22.0   
10           Jared Sullinger          Boston Celtics     7.0        C  24.0   
11             Isaiah Thomas          Boston Celtics     4.0       PG  27.0   
12               Evan Turner          Boston Celtics    11.0       SG  27.0   
13               James Young          Boston Celtics    13.0       SG  20.0   
14              Tyler Zeller          Boston Celtics    44.0        C  26.0   
15          Bojan Bogdanovic           Brooklyn Nets    44.0       SG  27.0   
16              Markel Brown           Brooklyn Nets    22.0       SG  24.0   
17           Wayne Ellington           Brooklyn Nets    21.0       SG  28.0   
18   Rondae Hollis-Jefferson           Brooklyn Nets    24.0       SG  21.0   
19              Jarrett Jack           Brooklyn Nets     2.0       PG  32.0   
20            Sergey Karasev           Brooklyn Nets    10.0       SG  22.0   
21           Sean Kilpatrick           Brooklyn Nets     6.0       SG  26.0   
22              Shane Larkin           Brooklyn Nets     0.0       PG  23.0   
23               Brook Lopez           Brooklyn Nets    11.0        C  28.0   
24          Chris McCullough           Brooklyn Nets     1.0       PF  21.0   
25               Willie Reed           Brooklyn Nets    33.0       PF  26.0   
26           Thomas Robinson           Brooklyn Nets    41.0       PF  25.0   
27                Henry Sims           Brooklyn Nets    14.0        C  26.0   
28              Donald Sloan           Brooklyn Nets    15.0       PG  28.0   
29            Thaddeus Young           Brooklyn Nets    30.0       PF  27.0   
..                       ...                     ...     ...      ...   ...   
427          Cliff Alexander  Portland Trail Blazers    34.0       PF  20.0   
428          Al-Farouq Aminu  Portland Trail Blazers     8.0       SF  25.0   
429          Pat Connaughton  Portland Trail Blazers     5.0       SG  23.0   
430             Allen Crabbe  Portland Trail Blazers    23.0       SG  24.0   
431                 Ed Davis  Portland Trail Blazers    17.0        C  27.0   
432         Maurice Harkless  Portland Trail Blazers     4.0       SF  23.0   
433         Gerald Henderson  Portland Trail Blazers     9.0       SG  28.0   
434              Chris Kaman  Portland Trail Blazers    35.0        C  34.0   
435           Meyers Leonard  Portland Trail Blazers    11.0       PF  24.0   
436           Damian Lillard  Portland Trail Blazers     0.0       PG  25.0   
437            C.J. McCollum  Portland Trail Blazers     3.0       SG  24.0   
438             Luis Montero  Portland Trail Blazers    44.0       SG  23.0   
439            Mason Plumlee  Portland Trail Blazers    24.0        C  26.0   
440            Brian Roberts  Portland Trail Blazers     2.0       PG  30.0   
441              Noah Vonleh  Portland Trail Blazers    21.0       PF  20.0   
442            Trevor Booker               Utah Jazz    33.0       PF  28.0   
443               Trey Burke               Utah Jazz     3.0       PG  23.0   
444               Alec Burks               Utah Jazz    10.0       SG  24.0   
445               Dante Exum               Utah Jazz    11.0       PG  20.

# value = 0m replaced all columns with values as 0 where it saw null values

# this will make sense in column salary but not in columns like name, college etc

# we call a series first to put our .fillna() method on just 1 column salary where we want to make null values to 0

In [60]:
basket_ball_team["Salary"]

0       7730337.0
1       6796117.0
2             NaN
3       1148640.0
4       5000000.0
5      12000000.0
6       1170960.0
7       2165160.0
8       1824360.0
9       3431040.0
10      2569260.0
11      6912869.0
12      3425510.0
13      1749840.0
14      2616975.0
15      3425510.0
16       845059.0
17      1500000.0
18      1335480.0
19      6300000.0
20      1599840.0
21       134215.0
22      1500000.0
23     19689000.0
24      1140240.0
25       947276.0
26       981348.0
27       947276.0
28       947276.0
29     11235955.0
          ...    
427      525093.0
428     8042895.0
429      625093.0
430      947276.0
431     6980802.0
432     2894059.0
433     6000000.0
434     5016000.0
435     3075880.0
436     4236287.0
437     2525160.0
438      525093.0
439     1415520.0
440     2854940.0
441     2637720.0
442     4775000.0
443     2658240.0
444     9463484.0
445     3777720.0
446    12000000.0
447     1175880.0
448    15409570.0
449     1348440.0
450     2050000.0
451      9

In [62]:
basket_ball_team["Salary"].fillna(value = 0, inplace = True)

In [63]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        0.0  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [64]:
basket_ball_team["College"].fillna(value= "Missing college info",inplace=True)

In [65]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

                College     Salary  
0                 Texas  7730337.0  
1             Marquette  6796117.0  
2     Boston University        0.0  
3         Georgia State  1148640.0  
4  Missing college info  5000000.0

In [66]:
# row 4 has missing college info

# ASTYPE() METHOD

# it allows data types of a series to convert from one to another

# data should not have any null values before we use this method so we remove any null values that exist

In [67]:
basket_ball_team= pd.read_csv("nba.csv")

In [68]:
basket_ball_team.tail()

Name       Team  Number Position   Age Height  Weight College  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   

        Salary  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0  
457        NaN

# last row has null values 

In [82]:
basket_ball_team =pd.read_csv("nba.csv").dropna(how='all')

In [83]:
basket_ball_team.tail()

Name       Team  Number Position   Age Height  Weight   College  \
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0    Kansas   

        Salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

In [84]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# column college and salary has NULL VALUES use method below to fill null values

In [89]:
basket_ball_team["Salary"].fillna(value=0,inplace=True)
basket_ball_team["College"].fillna(value="Missing college info",inplace=True)

In [90]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

                College     Salary  
0                 Texas  7730337.0  
1             Marquette  6796117.0  
2     Boston University        0.0  
3         Georgia State  1148640.0  
4  Missing college info  5000000.0

# for fillna() method you always extract the column as a SERIE before filling null values

# .astype() method

In [24]:
import pandas as pd

In [25]:
basket_ball_team = pd.read_csv("nba.csv")

In [26]:
basket_ball_team.tail()

Name       Team  Number Position   Age Height  Weight College  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   

        Salary  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0  
457        NaN

# we have to get rid of null values before using this method, below are the operations for that 

In [27]:
basket_ball_team = pd.read_csv("nba.csv").dropna(how='all')

In [28]:
basket_ball_team.tail()

Name       Team  Number Position   Age Height  Weight   College  \
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0    Kansas   

        Salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

In [29]:
basket_ball_team["Salary"].fillna(value=0,inplace=True)
basket_ball_team["College"].fillna(value='Missing Info',inplace=True)
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        0.0  
3      Georgia State  1148640.0  
4       Missing Info  5000000.0

In [30]:
basket_ball_team.dtypes

Name         object
Team         object
Number      float64
Position     object
Age         float64
Height       object
Weight      float64
College      object
Salary      float64
dtype: object

# object = string

In [31]:
basket_ball_team.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     457 non-null object
Salary      457 non-null float64
dtypes: float64(4), object(5)
memory usage: 35.7+ KB


# info tells us if we have any null values here

#  .astype() method LETS US CONVERT DATA TYPES - data type can be changed below :

# float to int

In [34]:
basket_ball_team["Salary"] = basket_ball_team["Salary"].astype("int")

In [35]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College   Salary  
0              Texas  7730337  
1          Marquette  6796117  
2  Boston University        0  
3      Georgia State  1148640  
4       Missing Info  5000000

In [36]:
basket_ball_team["Number"] = basket_ball_team["Number"].astype("int")

In [37]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics      99       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics      30       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics      28       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics       8       PF  29.0   6-10   231.0   

             College   Salary  
0              Texas  7730337  
1          Marquette  6796117  
2  Boston University        0  
3      Georgia State  1148640  
4       Missing Info  5000000

In [38]:
basket_ball_team["Age"] = basket_ball_team["Age"].astype("int")

In [39]:
basket_ball_team.head()

Name            Team  Number Position  Age Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25    6-2   180.0   
1    Jae Crowder  Boston Celtics      99       SF   25    6-6   235.0   
2   John Holland  Boston Celtics      30       SG   27    6-5   205.0   
3    R.J. Hunter  Boston Celtics      28       SG   22    6-5   185.0   
4  Jonas Jerebko  Boston Celtics       8       PF   29   6-10   231.0   

             College   Salary  
0              Texas  7730337  
1          Marquette  6796117  
2  Boston University        0  
3      Georgia State  1148640  
4       Missing Info  5000000

# to convert int data type back to float use the method below

In [40]:
basket_ball_team["Age"].astype("float")

0      25.0
1      25.0
2      27.0
3      22.0
4      29.0
5      29.0
6      21.0
7      25.0
8      22.0
9      22.0
10     24.0
11     27.0
12     27.0
13     20.0
14     26.0
15     27.0
16     24.0
17     28.0
18     21.0
19     32.0
20     22.0
21     26.0
22     23.0
23     28.0
24     21.0
25     26.0
26     25.0
27     26.0
28     28.0
29     27.0
       ... 
427    20.0
428    25.0
429    23.0
430    24.0
431    27.0
432    23.0
433    28.0
434    34.0
435    24.0
436    25.0
437    24.0
438    23.0
439    26.0
440    30.0
441    20.0
442    28.0
443    23.0
444    24.0
445    20.0
446    24.0
447    23.0
448    26.0
449    23.0
450    28.0
451    26.0
452    20.0
453    26.0
454    24.0
455    26.0
456    26.0
Name: Age, Length: 457, dtype: float64

# By specifying dtype="category" when constructing a Series:

#In [1]: s = pd.Series(["a","b","c","a"], dtype="category")

#In [2]: s
#Out[2]: 
#0    a
#1    b
#2    c
#3    a
#dtype: category
#Categories (3, object): [a, b, c]

In [43]:
basket_ball_team["Position"].nunique()

5

In [45]:
basket_ball_team["Position"] = basket_ball_team["Position"].astype("category")

In [46]:
basket_ball_team.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null int32
Position    457 non-null category
Age         457 non-null int32
Height      457 non-null object
Weight      457 non-null float64
College     457 non-null object
Salary      457 non-null int32
dtypes: category(1), float64(1), int32(3), object(4)
memory usage: 27.4+ KB


# reduces memory from 35.7 kb to 27.4 kb => makes sense when using large data sets it helps reduce memory usage

In [47]:
basket_ball_team["Team"] = basket_ball_team["Team"].astype("category")

In [48]:
basket_ball_team.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null category
Number      457 non-null int32
Position    457 non-null category
Age         457 non-null int32
Height      457 non-null object
Weight      457 non-null float64
College     457 non-null object
Salary      457 non-null int32
dtypes: category(2), float64(1), int32(3), object(3)
memory usage: 25.8+ KB


In [49]:
# memory usage further reduced 

# CATEGORY DATA TYPE = WHEN YOU HAVE LOT OF DUPLICATES WITHIN A COLUMN & RATHER THEN STORE THEM AS STRINGS YOU CAN STOR ETHEM AS CATEGORIES

# .SORT_VALUE() METHOD

In [50]:
basket_ball_team = pd.read_csv("nba.csv")

In [51]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [61]:
basket_ball_team.sort_values("Name",inplace=True)

basket_ball_team.sort_values("Age",ascending=True)

basket_ball_team.head()

Name                    Team  Number Position   Age Height  \
152    Aaron Brooks           Chicago Bulls     0.0       PG  31.0    6-0   
356    Aaron Gordon           Orlando Magic     0.0       PF  20.0    6-9   
328  Aaron Harrison       Charlotte Hornets     9.0       SG  21.0    6-6   
404   Adreian Payne  Minnesota Timberwolves    33.0       PF  25.0   6-10   
312      Al Horford           Atlanta Hawks    15.0        C  30.0   6-10   

     Weight         College      Salary  
152   161.0          Oregon   2250000.0  
356   220.0         Arizona   4171680.0  
328   210.0        Kentucky    525093.0  
404   237.0  Michigan State   1938840.0  
312   245.0         Florida  12000000.0

In [63]:
null_values = basket_ball_team.sort_values("Salary",ascending=False,na_position="first")

In [64]:
null_values.head()

Name                 Team  Number Position   Age Height  \
273  Alex Stepheson    Memphis Grizzlies    35.0       PF  28.0   6-10   
397    Axel Toupane       Denver Nuggets     6.0       SG  23.0    6-7   
350   Briante Weber           Miami Heat    12.0       PG  23.0    6-2   
171   Dahntay Jones  Cleveland Cavaliers    30.0       SG  35.0    6-6   
353   Dorell Wright           Miami Heat    11.0       SF  30.0    6-9   

     Weight                College  Salary  
273   270.0                    USC     NaN  
397   210.0                    NaN     NaN  
350   165.0  Virginia Commonwealth     NaN  
171   225.0                   Duke     NaN  
353   205.0                    NaN     NaN

# na_position = first gives all null values at the very top

# .SORT_VALUES() USING USING MULTIPLE COLUMNS

In [67]:
basket_ball_team = pd.read_csv("nba.csv")

In [68]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# sorts in alphabetical order = first the team and then name column

In [73]:
basket_ball_team.sort_values(["Team","Name"],ascending=[True,False]).head()

Name           Team  Number Position   Age Height  Weight  \
322    Walter Tavares  Atlanta Hawks    22.0        C  24.0    7-3   260.0   
310  Tim Hardaway Jr.  Atlanta Hawks    10.0       SG  24.0    6-6   205.0   
321    Tiago Splitter  Atlanta Hawks    11.0        C  31.0   6-11   245.0   
320   Thabo Sefolosha  Atlanta Hawks    25.0       SF  32.0    6-7   220.0   
315      Paul Millsap  Atlanta Hawks     4.0       PF  31.0    6-8   246.0   

            College      Salary  
322             NaN   1000000.0  
310        Michigan   1304520.0  
321             NaN   9756250.0  
320             NaN   4000000.0  
315  Louisiana Tech  18671659.0

# team = asc is True, Name = asc is false ----> this is how it relates to each column

# .SORT_INDEX() METHOD

In [119]:
basket_ball_team = pd.read_csv("nba.csv")

In [120]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [121]:
basket_ball_team.sort_values(["Name","Team","Age"],inplace=True)

In [122]:
basket_ball_team.head()

Name                    Team  Number Position   Age Height  \
152    Aaron Brooks           Chicago Bulls     0.0       PG  31.0    6-0   
356    Aaron Gordon           Orlando Magic     0.0       PF  20.0    6-9   
328  Aaron Harrison       Charlotte Hornets     9.0       SG  21.0    6-6   
404   Adreian Payne  Minnesota Timberwolves    33.0       PF  25.0   6-10   
312      Al Horford           Atlanta Hawks    15.0        C  30.0   6-10   

     Weight         College      Salary  
152   161.0          Oregon   2250000.0  
356   220.0         Arizona   4171680.0  
328   210.0        Kentucky    525093.0  
404   237.0  Michigan State   1938840.0  
312   245.0         Florida  12000000.0

In [123]:
basket_ball_team.sort_index().head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [124]:
basket_ball_team.sort_index(ascending=False).head()

Name       Team  Number Position   Age Height  Weight College  \
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   

        Salary  
457        NaN  
456   947276.0  
455  2900000.0  
454   900000.0  
453  2433333.0

# RANK VALUES WITH THE .RANK() METHOD

In [125]:
basket_ball_team = pd.read_csv("nba.csv")

In [126]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [127]:
columns_with_complete_null_values = basket_ball_team.dropna(how='all')

In [128]:
columns_with_complete_null_values.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

# dropna - how = 'all' removes column with ALL NULL VALUES

In [129]:
basket_ball_team["Salary"] = basket_ball_team["Salary"].fillna(value=0).astype("int")

In [130]:
basket_ball_team.head(3)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   

             College   Salary  
0              Texas  7730337  
1          Marquette  6796117  
2  Boston University        0

In [131]:
basket_ball_team["Salary Rank"] = basket_ball_team["Salary"].rank(ascending=False).astype("int")

In [132]:
basket_ball_team.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College   Salary  Salary Rank  
0              Texas  7730337           97  
1          Marquette  6796117          110  
2  Boston University        0          452  
3      Georgia State  1148640          322  
4                NaN  5000000          147

# ascending = False -> highest salary are going to possess the smaller ranks 

In [133]:
sorting_salary_column = basket_ball_team.sort_values(by = "Salary",ascending=False)

In [134]:
sorting_salary_column.head()

Name                 Team  Number Position   Age Height  \
109      Kobe Bryant   Los Angeles Lakers    24.0       SF  37.0    6-6   
169     LeBron James  Cleveland Cavaliers    23.0       SF  31.0    6-8   
33   Carmelo Anthony      New York Knicks     7.0       SF  32.0    6-8   
251    Dwight Howard      Houston Rockets    12.0        C  30.0   6-11   
339       Chris Bosh           Miami Heat     1.0       PF  32.0   6-11   

     Weight       College    Salary  Salary Rank  
109   212.0           NaN  25000000            1  
169   250.0           NaN  22970500            2  
33    240.0      Syracuse  22875000            3  
251   265.0           NaN  22359364            4  
339   235.0  Georgia Tech  22192730            5

# highest salary has rank 1